components/formats-gpl/src/loci/formats/out/ZarrWriter.java

```diff
@@ -0,0 +1,209 @@
+package loci.formats.out;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+import java.nio.ShortBuffer;
+import java.util.UUID;
+
+import loci.common.Constants;
+import loci.common.Location;
+import loci.common.services.DependencyException;
+import loci.common.services.ServiceException;
+import loci.common.services.ServiceFactory;
+import loci.common.xml.XMLTools;
+import loci.formats.FormatException;
+import loci.formats.FormatTools;
+import loci.formats.FormatWriter;
+import loci.formats.MetadataTools;
+import loci.formats.meta.MetadataRetrieve;
+import loci.formats.ome.OMEXMLMetadata;
+import loci.formats.services.ZarrServiceImpl;
+import ome.xml.meta.OMEXMLMetadataRoot;
+import loci.formats.services.OMEXMLService;
+import loci.formats.services.ZarrService;
+
+
+public class ZarrWriter extends FormatWriter {
+
+  // -- Fields --
+  private ZarrService zarrService = null;
+  private final int CHUNK_DEFAULT = 32;
+  private OMEXMLService service;
+  private OMEXMLMetadata omeMeta;
+  
+  
+  public ZarrWriter() {
+    this("Zarr", new String[] {"zarr"});
+  }
+ 
+  public ZarrWriter(String format, String[] exts) {
+    super(format, exts);
+  }
+  
+  
+  // -- IFormatWriter API methods --
+
+  /**
+   * @see loci.formats.IFormatWriter#saveBytes(int, byte[], int, int, int, int)
+   */
+  @Override
+  public void saveBytes(int no, byte[] buf, int x, int y, int w, int h)
+    throws FormatException, IOException
+  {
+    if(zarrService != null) {
+      checkParams(no, buf, x, y, w, h);
+      MetadataRetrieve meta = getMetadataRetrieve();
+      MetadataTools.verifyMinimumPopulated(meta, series);
+      
+      int zSize = meta.getPixelsSizeZ(series).getValue().intValue();
+      int cSize = meta.getPixelsSizeC(series).getValue().intValue();
+      int tSize = meta.getPixelsSizeT(series).getValue().intValue();
+      String order = meta.getPixelsDimensionOrder(series).toString();
+      int imageCount = getPlaneCount();
+      int[] coordinates = FormatTools.getZCTCoords(order, zSize, cSize, tSize, imageCount, no);
+      int [] shape = {w, h, 1, 1, 1};
+
+      int type = FormatTools.pixelTypeFromString(meta.getPixelsType(0).toString());
+      int bpp = FormatTools.getBytesPerPixel(type);
+      
+      int [] offsets = {x, y, coordinates[0], coordinates[1], coordinates[2]};
+      if (bpp==1) {
+        zarrService.saveBytes(buf, shape, offsets);
+      }
+      else if (bpp==2) {
+        ByteBuffer bb = ByteBuffer.wrap(buf);
+        ShortBuffer sb = bb.asShortBuffer();
+        short[] image = new short[w * h];
+        for (int x_i=0; x_i < w; x_i++) {
+            for (int y_i=0; y_i < h; y_i++) {
+                short value = sb.get(y_i*w + x_i);
+                image[y_i*w + x_i] = value;
+            }
+        }
```

Comment:
Since the source and destination indexes are the same, is it necessary to iterate over each pixel or could this be replaced by ```sb.get(image)```?